pkg/intermediate/aggregate.go

```diff
@@ -148,31 +172,93 @@ func (a *AggregationProcess) ForAllRecordsDo(callback FlowKeyRecordMapCallBack)
 	return nil
 }
 
-// GetLastUpdatedTimeOfFlow provides the last updated time in the format of IPFIX
-// field "flowEndSeconds".
-func (a *AggregationProcess) GetLastUpdatedTimeOfFlow(flowKey FlowKey) (uint32, error) {
+func (a *AggregationProcess) deleteFlowKeyFromMap(flowKey FlowKey) error {
 	a.mutex.Lock()
 	defer a.mutex.Unlock()
-	record, exists := a.flowKeyRecordMap[flowKey]
+	return a.deleteFlowKeyFromMapWithoutLock(flowKey)
+}
+
+func (a *AggregationProcess) deleteFlowKeyFromMapWithoutLock(flowKey FlowKey) error {
+	_, exists := a.flowKeyRecordMap[flowKey]
 	if !exists {
-		return 0, fmt.Errorf("flow key is not present in the map")
+		return fmt.Errorf("flow key %v is not present in the map", flowKey)
 	}
-	flowEndField, exists := record.Record.GetInfoElementWithValue("flowEndSeconds")
-	if exists {
-		return flowEndField.Value.(uint32), nil
-	} else {
-		return 0, fmt.Errorf("flowEndSeconds field is not present in the record")
+	delete(a.flowKeyRecordMap, flowKey)
+	return nil
+}
+
+// GetExpiryFromExpirePriorityQueue returns the earliest timestamp (active expiry
+// or inactive expiry) from expire priority queue.
+func (a *AggregationProcess) GetExpiryFromExpirePriorityQueue() time.Duration {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	currTime := time.Now()
+	if a.expirePriorityQueue.Len() > 0 {
+		// Get the minExpireTime of the top item in expirePriorityQueue.
+		expiryDuration := MinExpiryTime + a.expirePriorityQueue.minExpireTime(0).Sub(currTime)
+		if expiryDuration < 0 {
+			return MinExpiryTime
```

Comment:
You may no longer need to add MinExpiryTime on previous line?

Reply:
MinExpiryTime in the previous line is to process multiple records simultaneously if their expiry times are very close to each other (<100ms).

Reply:
Got it. Maybe then doing like this?

    expiryDuration := a.expirePriorityQueue.minExpireTime(0).Sub(currTime)
    if expiryDuration < MinExpiryTime  {
        return MinExpiryTime
    }

Reply:
Sorry not quite understand the scenario.
I thought we just want to make sure this function will return at least 100ms for each call (assume non-empty queue case)?
Did you mean there are some cases, you want it to return more than 100ms?

Reply:
Yes, that is correct. This is to make sure that we do not process single expired record, if the next record expiry time instance is lower than 100ms.

Reply:
Got it. But in the statement

    expiryDuration := MinExpiryTime + a.expirePriorityQueue.minExpireTime(0).Sub(currTime),

if a.expirePriorityQueue.minExpireTime(0).Sub(currTime) = -20ms,
then this function will return 80ms to the caller, which is less than 100ms.
Will this be OK?

Reply:
Yes that will be fine. It's a best effort to have some minimum delay to avoid single flow record processing scenarios.

Reply:
>Is index 1 in your reply the first or second item in the queue
Yes that is second item and not top item.

>So you're trying to avoid the returned duration covers only one item?
Yes for cases, where second item expiry item is close to the first one. Think of batch processing. Just a heuristic-based small enhancement.

Reply:
OK, got it now.
I thought configuring a reasonable MinExpiryTime is good enough since there could be always a item very close to the chosen MinExpiryTime.
Thanks.